pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/tpch/generator/TPCHQueryGeneratorV2.java

```diff
@@ -0,0 +1,553 @@
+    }
+    for (Table value : tables) {
+      querySkeleton.addTable(value.getTableName());
+    }
+
+    if (predicates.size() > 0) {
+      for (String predicate : predicates) {
+        querySkeleton.addPredicate(predicate);
+      }
+    }
+
+    if (includePredicates) {
+      for (Table table : tables) {
+        List<String> preds = getRandomPredicates(table);
+        preds.forEach(querySkeleton::addPredicate);
+      }
+    }
+
+    if (includeOrderBy) {
+      for (Table table : tables) {
+        getRandomOrderBys(table).forEach(querySkeleton::addOrderByColumn);
+      }
+    }
+
+    return querySkeleton.toString();
+  }
+
+  public String selectionGroupByMultiJoin(boolean includePredicates, boolean includeOrderBy) {
+    QuerySkeleton querySkeleton = new QuerySkeleton();
+
+    List<String> predicates = new ArrayList<>();
+    List<Table> tables = new ArrayList<>();
+    Set<String> tableNames = new HashSet<>();
+
+    Random random = new Random();
+
+    // Start off with a random table with related tables
+    while (true) {
+      Table t1 = getRandomTable();
+      if (t1.getRelatedTables().size() > 0) {
+        tables.add(t1);
+        tableNames.add(t1.getTableName());
+        break;
+      }
+    }
+
+    // Add more tables
+    while (random.nextInt() % 8 != 0) {
+      int tableToAddIdx = random.nextInt(tables.size());
+      RelatedTable relatedTable = tables.get(tableToAddIdx).getRelatedTables()
+          .get(random.nextInt(tables.get(tableToAddIdx).getRelatedTables().size()));
+      if (!tableNames.contains(relatedTable.getForeignTableName())) {
+        tableNames.add(relatedTable.getForeignTableName());
+        tables.add(TPCHQueryGeneratorV2.TABLES_MAP.get(relatedTable.getForeignTableName()));
+        predicates.add(relatedTable.getLocalTableKey() + "=" + relatedTable.getForeignTableKey());
+      }
+    }
+
+    Map<String, List<String>> tableWiseGroupByCols = new HashMap<>();
+    for (Table value : tables) {
+      Pair<List<String>, List<String>> groupByAndAggregates = getGroupByAndAggregates(value);
+      groupByAndAggregates.getLeft().forEach(querySkeleton::addProjection);
+      groupByAndAggregates.getRight().forEach(querySkeleton::addGroupByColumn);
+      tableWiseGroupByCols.put(value.getTableName(), groupByAndAggregates.getRight());
+    }
+    for (Table table : tables) {
+      querySkeleton.addTable(table.getTableName());
+    }
+    predicates.forEach(querySkeleton::addPredicate);
+
+    if (includePredicates) {
+      for (Table table : tables) {
+        List<String> preds = getRandomPredicates(table);
+        preds.forEach(querySkeleton::addPredicate);
+      }
+    }
+
+    if (includeOrderBy) {
+      for (Table table : tables) {
+        getRandomOrderBys(table, tableWiseGroupByCols.get(table.getTableName())).forEach(
+            querySkeleton::addOrderByColumn);
+      }
+    }
+
+    return querySkeleton.toString();
+  }
+
+  public String generateRandomQuery() {
```

Comment:
did we come up with this structure from some reference database or this is purely just coming from us?

- it seems like we didn't have distinct (or group by without agg) examples?
- i saw group-by after join, is there any join after group-by queries? 
- window function is not covered i suppose

did we have a follow up on what needs to be added next?

Reply:
@walterddr , this was entirely coming as an extension of https://github.com/apache/pinot/blob/master/pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/QueryGenerator.java#L67, but with support for joins, aggregations, group bys etc.

I can work with you to identify follow ups for this. I think majorly we're missing nested queries, exhaustive set of aggregations, window function etc.